GVFS/GVFS/CommandLine/DehydrateVerb.cs

```diff
@@ -145,31 +164,139 @@ protected override void Execute(GVFSEnlistment enlistment)
+                {
+                    string[] folders = this.Folders.Split(new[] { ";" }, StringSplitOptions.RemoveEmptyEntries);
+
+                    if (folders.Length > 0)
                     {
-                        // Converting the src folder to partial must be the final step before mount
-                        this.PrepareSrcFolder(tracer, enlistment);
-                        this.Mount(tracer);
+                        if (cleanStatus)
+                        {
+                            this.DehydrateFolders(tracer, enlistment, backupRoot, folders);
+                        }
+                        else
+                        {
+                            this.ReportErrorAndExit("Must have a clean git status to dehydrate folders.");
+                        }
+                    }
+                    else
+                    {
+                        this.ReportErrorAndExit("No folders to dehydrate.");
+                    }
+                }
+            }
+        }
+
+        private void DehydrateFolders(JsonTracer tracer, GVFSEnlistment enlistment, string backupRoot, string[] folders)
+        {
+            List<string> foldersToDehydrate = new List<string>();
+
+            if (!this.ShowStatusWhileRunning(
+                () =>
+                {
+                    string ioError;
+                    foreach (string folder in folders)
+                    {
+                        string fullPath = Path.Combine(enlistment.WorkingDirectoryRoot, folder);
+                        string backupPath = Path.Combine(backupRoot, folder);
+                        if (this.fileSystem.DirectoryExists(fullPath))
+                        {
+                            if (!this.TryIO(tracer, () => this.fileSystem.CopyDirectoryRecursive(fullPath, backupPath), $"Backing up {folder} to {backupPath}", out ioError) ||
+                                !this.TryIO(tracer, () => this.fileSystem.DeleteDirectory(fullPath), $"Deleting {fullPath}", out ioError))
+                            {
+                                this.WriteMessage(tracer, $"Backup failed for {folder} and will not be dehydrated. {ioError}");
+                                this.WriteMessage(tracer, $"Make sure there aren't any applications accessing the folder and try again.");
+                            }
+                            else
+                            {
+                                foldersToDehydrate.Add(folder);
+                            }
+                        }
+                        else
+                        {
+                            this.WriteMessage(tracer, $"{folder} did not exist to dehydrate.");
+                        }
+                    }
+
+                    return true;
+                },
+                "Backing up folders"))
+            {
+                this.ReportErrorAndExit(tracer, "Dehydrate for folders failed.");
+            }
+
+            this.Mount(tracer);

```

Comment:
If this call to `Mount` were to fail, what would a user do to repair their repo and try the dehydrate again?  

I'm wondering if the dehydrate verb should send over the dehydrate message even when the folder does not exist on disk.  Would doing so harm anything?

Reply:
>If this call to Mount were to fail, what would a user do to repair their repo and try the dehydrate again?

I guess that would depend on why the mount failed.  Did you have some thoughts on what would cause this?

>I'm wondering if the dehydrate verb should send over the dehydrate message even when the folder does not exist on disk. Would doing so harm anything?

Thinking this through and I don't think it would harm anything.  It would remove modified paths and placeholders for the folder and if it was included create the placeholder on Mac but that shouldn't be the case because there was a clean `git status` so it shouldn't be in the index as included but not on disk.  I will try it out and see.

Reply:
I think I understand what you were trying to point out that if the mount fails but we have already deleted the folders that were passed in, if the user runs dehydrate again we would still want to send the dehydrate message to clean up those folders which makes sense.